File: src/testing/command-conformance.ts
import type { CommandContext, CommandRegistry, CommandSpec } from '../commands/index.js';
import type { CommandResult } from '../types.js';
import { textDecoder, textEncoder } from '../types.js';

export interface CommandConformanceOptions {
  registry: CommandRegistry;
  makeCtx: (overrides?: Partial<CommandContext>) => CommandContext;
  strictMetadata?: boolean;
  rejectionStdin?: Uint8Array;
}

export interface CommandConformanceCase {
  commandName: string;
  name: string;
  run: () => Promise<void>;
}

const NO_STDIN = new Uint8Array();

export function createCommandConformanceCases(options: CommandConformanceOptions): CommandConformanceCase[] {
  const registry = options.registry;
  const strictMetadata = options.strictMetadata === true;
  const cases: CommandConformanceCase[] = [];

  for (const spec of registry.all()) {
    const sampleArgs = sampleArgsFor(spec);

    cases.push(
      createCase(spec, 'has valid spec metadata', async function (): Promise<void> {
        assertSpecMetadata(spec, sampleArgs, strictMetadata);
      }),
    );

    cases.push(
      createCase(spec, 'representative call never throws', async function (): Promise<void> {
        await invokeSpec(options, spec, sampleArgs);
      }),
    );

    cases.push(
      createCase(spec, 'error output references command name', async function (): Promise<void> {
        const result = await invokeSpec(options, spec, []);
        if (result.exitCode !== 0) {
          assertIncludes(
            result.stderr,
            spec.name,
            `error output should reference "${spec.name}": got "${result.stderr}"`,
          );
        }
      }),
    );

    if (registry.has('help')) {
      cases.push(
        createCase(spec, 'help entry exists', async function (): Promise<void> {
          const result = await invokeHelp(options, spec.name);
          assertEqual(result.exitCode, 0, `help ${spec.name} should exit successfully`);
          assertIncludes(
            stdoutText(result),
            spec.summary,
            `help ${spec.name} should include the command summary`,
          );
        }),
      );
    }

    if (spec.acceptsStdin === false) {
      cases.push(
        createCase(spec, 'rejects stdin', async function (): Promise<void> {
          const stdin = options.rejectionStdin ?? textEncoder.encode('unwanted');
          const result = await invokeSpec(options, spec, sampleArgs, stdin);
          assertEqual(result.exitCode, 1, `${spec.name} should reject stdin`);
          assertMatch(
            result.stderr,
            /does not accept stdin/,
            `${spec.name} should explain the stdin rejection`,
          );
        }),
      );
    }

    if ((spec.minArgs ?? 0) > 0) {
      cases.push(
        createCase(spec, `requires at least ${spec.minArgs} arg(s)`, async function (): Promise<void> {
          const result = await invokeSpec(options, spec, tooFewArgsFor(spec));
          assertCondition(result.exitCode !== 0, `${spec.name} should reject too few arguments`);
        }),
      );
    }

    if (spec.maxArgs !== undefined) {
      cases.push(
        createCase(spec, `rejects more than ${spec.maxArgs} arg(s)`, async function (): Promise<void> {
          const result = await invokeSpec(options, spec, tooManyArgsFor(spec));
          assertCondition(result.exitCode !== 0, `${spec.name} should reject too many arguments`);
        }),
      );
    }

    if (spec.requiresAdapter !== undefined) {
      cases.push(
        createCase(
          spec,
          `reports missing ${String(spec.requiresAdapter)} adapter`,
          async function (): Promise<void> {
            const result = await invokeSpec(options, spec, adapterArgsFor(spec, sampleArgs), NO_STDIN, {
              adapters: {},
            });
            assertEqual(result.exitCode, 1, `${spec.name} should reject missing adapters`);
            assertMatch(
              result.stderr,
              new RegExp(
                `${escapeRegExp(spec.name)}: no ${escapeRegExp(String(spec.requiresAdapter))} adapter configured`,
              ),
              `${spec.name} should explain the missing adapter`,
            );
          },
        ),
      );
    }
  }

  return cases;
}

function createCase(spec: CommandSpec, name: string, run: () => Promise<void>): CommandConformanceCase {
  return {
    commandName: spec.name,
    name,
    run,
  };
}

function sampleArgsFor(spec: CommandSpec): string[] {
  return [...(spec.conformanceArgs ?? [])];
}

function adapterArgsFor(spec: CommandSpec, sampleArgs: string[]): string[] {
  if (sampleArgs.length > 0) {
    return [...sampleArgs];
  }

  const count = Math.max(spec.minArgs ?? 0, 1);
  return Array.from({ length: count }, function (_, index) {
    return `arg${index}`;
  });
}

function tooFewArgsFor(spec: CommandSpec): string[] {
  const minArgs = spec.minArgs ?? 0;
  return Array.from({ length: minArgs - 1 }, function (_, index) {
    return `arg${index}`;
  });
}

function tooManyArgsFor(spec: CommandSpec): string[] {
  const maxArgs = spec.maxArgs ?? 0;
  return Array.from({ length: maxArgs + 1 }, function (_, index) {
    return `arg${index}`;
  });
}

function assertSpecMetadata(spec: CommandSpec, sampleArgs: string[], strictMetadata: boolean): void {
  assertNonEmpty(spec.name, 'command name');
  assertNonEmpty(spec.summary, `${spec.name} summary`);
  assertNonEmpty(spec.usage, `${spec.name} usage`);
  assertNonEmpty(spec.details, `${spec.name} details`);
  assertCondition(typeof spec.handler === 'function', `${spec.name} handler must be a function`);

  if (strictMetadata) {
    assertCondition(typeof spec.acceptsStdin === 'boolean', `${spec.name} should declare acceptsStdin`);
    assertCondition(typeof spec.minArgs === 'number', `${spec.name} should declare minArgs`);
    assertCondition(Array.isArray(spec.conformanceArgs), `${spec.name} should declare conformanceArgs`);
  } else {
    if (spec.acceptsStdin !== undefined) {
      assertCondition(typeof spec.acceptsStdin === 'boolean', `${spec.name} acceptsStdin must be boolean`);
    }
    if (spec.minArgs !== undefined) {
      assertCondition(typeof spec.minArgs === 'number', `${spec.name} minArgs must be numeric`);
    }
    if (spec.conformanceArgs !== undefined) {
      assertCondition(Array.isArray(spec.conformanceArgs), `${spec.name} conformanceArgs must be an array`);
    }
  }

  if (spec.maxArgs !== undefined) {
    assertCondition(typeof spec.maxArgs === 'number', `${spec.name} maxArgs must be numeric`);
    assertCondition(spec.maxArgs >= (spec.minArgs ?? 0), `${spec.name} has maxArgs < minArgs`);
  }
  if (spec.minArgs !== undefined) {
    assertCondition(sampleArgs.length >= spec.minArgs, `${spec.name} conformanceArgs should satisfy minArgs`);
  }
  if (spec.maxArgs !== undefined) {
    assertCondition(sampleArgs.length <= spec.maxArgs, `${spec.name} conformanceArgs should satisfy maxArgs`);
  }
}

async function invokeHelp(options: CommandConformanceOptions, commandName: string): Promise<CommandResult> {
  const ctx = options.makeCtx({ registry: options.registry });
  const help = ctx.registry.get('help');
  if (!help) {
    throw new Error('help command is not registered');
  }
  return help.handler(ctx, [commandName], NO_STDIN);
}

async function invokeSpec(
  options: CommandConformanceOptions,
  spec: CommandSpec,
  args: string[],
  stdin: Uint8Array = NO_STDIN,
  ctxOverrides: Partial<CommandContext> = {},
): Promise<CommandResult> {
  const ctx = options.makeCtx({ registry: options.registry, ...ctxOverrides });
  return spec.handler(ctx, [...args], stdin);
}

function stdoutText(result: CommandResult): string {
  return textDecoder.decode(result.stdout);
}

function assertCondition(condition: boolean, message: string): void {
  if (!condition) {
    throw new Error(message);
  }
}

function assertEqual(actual: unknown, expected: unknown, message: string): void {
  if (actual !== expected) {
    throw new Error(`${message}. Expected ${String(expected)}, got ${String(actual)}`);
  }
}

function assertIncludes(actual: string, expected: string, message: string): void {
  if (!actual.includes(expected)) {
    throw new Error(`${message}. Expected "${actual}" to include "${expected}"`);
  }
}

function assertMatch(actual: string, pattern: RegExp, message: string): void {
  if (!pattern.test(actual)) {
    throw new Error(`${message}. Expected "${actual}" to match ${pattern}`);
  }
}

function assertNonEmpty(value: string, label: string): void {
  assertCondition(value.trim().length > 0, `${label} must be non-empty`);
}

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

File: python/src/onetool/testing/command_conformance.py
from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
import re

from ..commands import CommandContext, CommandRegistry, CommandSpec, maybe_await
from ..types import CommandResult, ToolAdapters
from .command_harness import (
    NO_STDIN,
    CreateTestCommandContextOptions,
    create_test_command_context,
    get_registered_command,
    stdout_text,
)


@dataclass(frozen=True, slots=True)
class CommandConformanceOptions:
    registry: CommandRegistry
    make_ctx: Callable[[CreateTestCommandContextOptions | None], CommandContext]
    strict_metadata: bool = False
    rejection_stdin: bytes = b"unwanted"


@dataclass(frozen=True, slots=True)
class CommandConformanceCase:
    command_name: str
    name: str
    run: Callable[[], object]


def create_command_conformance_cases(options: CommandConformanceOptions) -> list[CommandConformanceCase]:
    cases: list[CommandConformanceCase] = []
    registry = options.registry

    for spec in registry.all():
        sample_args = list(spec.conformance_args)

        cases.append(
            _create_case(
                spec,
                "has valid spec metadata",
                lambda spec=spec, sample_args=sample_args: _assert_spec_metadata(
                    spec,
                    sample_args,
                    options.strict_metadata,
                ),
            )
        )
        cases.append(
            _create_case(
                spec,
                "representative call never throws",
                lambda spec=spec, sample_args=sample_args: _invoke_spec(options, spec, sample_args),
            )
        )
        cases.append(
            _create_case(
                spec,
                "error output references command name",
                lambda spec=spec: _assert_error_output_references_command(options, spec),
            )
        )

        if registry.has("help"):
            cases.append(
                _create_case(
                    spec,
                    "help entry exists",
                    lambda spec=spec: _assert_help_entry_exists(options, spec.name),
                )
            )

        if spec.accepts_stdin is False:
            cases.append(
                _create_case(
                    spec,
                    "rejects stdin",
                    lambda spec=spec, sample_args=sample_args: _assert_stdin_rejected(
                        options,
                        spec,
                        sample_args,
                    ),
                )
            )

        if (spec.min_args or 0) > 0:
            cases.append(
                _create_case(
                    spec,
                    f"requires at least {spec.min_args} arg(s)",
                    lambda spec=spec: _assert_too_few_args_rejected(options, spec),
                )
            )

        if spec.max_args is not None:
            cases.append(
                _create_case(
                    spec,
                    f"rejects more than {spec.max_args} arg(s)",
                    lambda spec=spec: _assert_too_many_args_rejected(options, spec),
                )
            )

        if spec.requires_adapter is not None:
            cases.append(
                _create_case(
                    spec,
                    f"reports missing {spec.requires_adapter} adapter",
                    lambda spec=spec, sample_args=sample_args: _assert_missing_adapter_rejected(
                        options,
                        spec,
                        sample_args,
                    ),
                )
            )

    return cases


def _create_case(spec: CommandSpec, name: str, run: Callable[[], object]) -> CommandConformanceCase:
    return CommandConformanceCase(
        command_name=spec.name,
        name=name,
        run=run,
    )


async def _assert_error_output_references_command(
    options: CommandConformanceOptions,
    spec: CommandSpec,
) -> None:
    result = await _invoke_spec(options, spec, [])
    if result.exit_code != 0:
        _assert_includes(
            result.stderr,
            spec.name,
            f'error output should reference "{spec.name}"',
        )


async def _assert_help_entry_exists(options: CommandConformanceOptions, command_name: str) -> None:
    ctx = options.make_ctx(CreateTestCommandContextOptions(registry=options.registry))
    help_spec = ctx.registry.get("help")
    if help_spec is None:
        raise ValueError("help command is not registered")
    result = await maybe_await(help_spec.handler(ctx, [command_name], NO_STDIN))
    _assert_equal(result.exit_code, 0, f"help {command_name} should exit successfully")
    _assert_includes(
        stdout_text(result),
        get_registered_command(command_name, options.registry).summary,
        f"help {command_name} should include the command summary",
    )


async def _assert_stdin_rejected(
    options: CommandConformanceOptions,
    spec: CommandSpec,
    sample_args: list[str],
) -> None:
    result = await _invoke_spec(options, spec, sample_args, options.rejection_stdin)
    _assert_equal(result.exit_code, 1, f"{spec.name} should reject stdin")
    _assert_match(result.stderr, r"does not accept stdin", f"{spec.name} should explain the stdin rejection")


async def _assert_too_few_args_rejected(options: CommandConformanceOptions, spec: CommandSpec) -> None:
    result = await _invoke_spec(options, spec, _too_few_args_for(spec))
    _assert_condition(result.exit_code != 0, f"{spec.name} should reject too few arguments")


async def _assert_too_many_args_rejected(options: CommandConformanceOptions, spec: CommandSpec) -> None:
    result = await _invoke_spec(options, spec, _too_many_args_for(spec))
    _assert_condition(result.exit_code != 0, f"{spec.name} should reject too many arguments")


async def _assert_missing_adapter_rejected(
    options: CommandConformanceOptions,
    spec: CommandSpec,
    sample_args: list[str],
) -> None:
    result = await _invoke_spec(
        options,
        spec,
        _adapter_args_for(spec, sample_args),
        ctx_options=CreateTestCommandContextOptions(
            registry=options.registry,
            adapters=ToolAdapters(),
        ),
    )
    _assert_equal(result.exit_code, 1, f"{spec.name} should reject missing adapters")
    _assert_match(
        result.stderr,
        rf"{re.escape(spec.name)}: no {re.escape(str(spec.requires_adapter))} adapter configured",
        f"{spec.name} should explain the missing adapter",
    )


async def _invoke_spec(
    options: CommandConformanceOptions,
    spec: CommandSpec,
    args: list[str],
    stdin: bytes = NO_STDIN,
    ctx_options: CreateTestCommandContextOptions | None = None,
) -> CommandResult:
    ctx = options.make_ctx(ctx_options or CreateTestCommandContextOptions(registry=options.registry))
    return await maybe_await(spec.handler(ctx, list(args), stdin))


def _assert_spec_metadata(spec: CommandSpec, sample_args: list[str], strict_metadata: bool) -> None:
    _assert_non_empty(spec.name, "command name")
    _assert_non_empty(spec.summary, f"{spec.name} summary")
    _assert_non_empty(spec.usage, f"{spec.name} usage")
    _assert_non_empty(spec.details, f"{spec.name} details")
    _assert_condition(callable(spec.handler), f"{spec.name} handler must be callable")

    if strict_metadata:
        _assert_condition(spec.accepts_stdin is not None, f"{spec.name} should declare accepts_stdin")
        _assert_condition(spec.min_args is not None, f"{spec.name} should declare min_args")
    elif spec.min_args is not None:
        _assert_condition(isinstance(spec.min_args, int), f"{spec.name} min_args must be numeric")

    if spec.accepts_stdin is not None:
        _assert_condition(isinstance(spec.accepts_stdin, bool), f"{spec.name} accepts_stdin must be boolean")
    if spec.max_args is not None:
        _assert_condition(isinstance(spec.max_args, int), f"{spec.name} max_args must be numeric")
        _assert_condition(spec.max_args >= (spec.min_args or 0), f"{spec.name} has max_args < min_args")
    if spec.min_args is not None:
        _assert_condition(len(sample_args) >= spec.min_args, f"{spec.name} conformance_args should satisfy min_args")
    if spec.max_args is not None:
        _assert_condition(len(sample_args) <= spec.max_args, f"{spec.name} conformance_args should satisfy max_args")


def _adapter_args_for(spec: CommandSpec, sample_args: list[str]) -> list[str]:
    if sample_args:
        return list(sample_args)
    count = max(spec.min_args or 0, 1)
    return [f"arg{index}" for index in range(count)]


def _too_few_args_for(spec: CommandSpec) -> list[str]:
    count = max((spec.min_args or 0) - 1, 0)
    return [f"arg{index}" for index in range(count)]


def _too_many_args_for(spec: CommandSpec) -> list[str]:
    count = (spec.max_args or 0) + 1
    return [f"arg{index}" for index in range(count)]


def _assert_condition(condition: bool, message: str) -> None:
    if not condition:
        raise AssertionError(message)


def _assert_equal(actual: object, expected: object, message: str) -> None:
    if actual != expected:
        raise AssertionError(f"{message}. Expected {expected!r}, got {actual!r}")


def _assert_includes(actual: str, expected: str, message: str) -> None:
    if expected not in actual:
        raise AssertionError(f'{message}. Expected "{actual}" to include "{expected}"')


def _assert_match(actual: str, pattern: str, message: str) -> None:
    if re.search(pattern, actual) is None:
        raise AssertionError(f'{message}. Expected "{actual}" to match /{pattern}/')


def _assert_non_empty(value: str, label: str) -> None:
    _assert_condition(value.strip() != "", f"{label} must be non-empty")
